File: internal/components/CheckboxGroup.js
(function () {
  const checkboxGroup = document.querySelector(".checkbox-group");
  const checkboxOptions = checkboxGroup.querySelector(".checkbox-options");
  const hiddenInput = checkboxGroup.querySelector(".hidden-input");
  const options = [];

  checkboxOptions.addEventListener("change", (e) => {
    if (e.target.tagName === "INPUT") {
      const option = e.target.value;
      if (e.target.checked) {
        options.push(option);
        hiddenInput.value = JSON.stringify(options);
        if (e.target.hasAttribute("checked")) {
          e.target.setAttribute("checked", "true");
        } else {
          e.target.setAttribute("checked", "true");
        }
      } else {
        options.splice(options.indexOf(option), 1);
        hiddenInput.value = JSON.stringify(options);
        if (e.target.hasAttribute("checked")) {
          e.target.setAttribute("checked", "false");
        } else {
          e.target.removeAttribute("checked");
        }
      }
    }
  });

  document.addEventListener("DOMContentLoaded", () => {
    const checkboxOption = checkboxOptions.querySelectorAll("input");
    checkboxOption.forEach((option) => {
      if (option.checked) {
        options.push(option.value);
        hiddenInput.value = JSON.stringify(options);
      }
    });
  });
})();

File: internal/components/CheckboxGroup.go
package components

import (
	g "maragu.dev/gomponents"
	c "maragu.dev/gomponents/components"
	h "maragu.dev/gomponents/html"
)

type CheckboxOption struct {
	Value string
	Label string
}

type CheckboxGroupProps struct {
	Name    string
	Label   string
	Options []CheckboxOption
	Value   []string
	Classes c.Classes
}

func CheckboxGroup(p *CheckboxGroupProps) g.Node {
	if p.Options == nil {
		p.Options = []CheckboxOption{}
	}

	if p.Name == "" {
		p.Name = "checkbox-group"
	}

	if p.Label == "" {
		p.Label = "Checkbox Group"
	}

	if p.Classes == nil {
		p.Classes = c.Classes{}
	}

	p.Classes["checkbox-group"] = true

	return h.Div(
		p.Classes,
		h.Label(h.For(p.Name), g.Text(p.Label)),
		h.Div(
			h.Class("checkbox-options"),
			g.Group(g.Map(p.Options, func(option CheckboxOption) g.Node {
				checked := false
				for _, value := range p.Value {
					if value == option.Value {
						checked = true
						break
					}
				}
				return h.Div(
					h.Class("checkbox-option"),
					Checkbox(&CheckboxProps{
						Name:    p.Name,
						Value:   option.Value,
						Checked: checked,
					}),
					g.Text(option.Label),
				)
			})),
		),
	)
}
